Library/Homebrew/livecheck/strategy/crates.rb

```diff
@@ -0,0 +1,111 @@
+# typed: true
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module Strategy
+      # The {Crates} strategy identifies versions of a crates.io package by
+      # checking the information from the `versions` API endpoint.
+      #
+      # Crate URLs have the following format:
+      #   `https://static.crates.io/crates/example/example-1.2.3.crate`
+      #
+      # The default regex identifies versions like `1.2.3`/`v1.2.3` from the
+      # version `num` field. This is a common version format but a different
+      # regex can be provided in a `livecheck` block to override the default
+      # if a package uses a different format (e.g. `1.2.3d`, `1.2.3-4`, etc.).
+      #
+      # @api public
+      class Crates
+        # The default regex used to identify versions when a regex isn't
+        # provided.
+        DEFAULT_REGEX = /^v?(\d+(?:\.\d+)+)$/i
+
+        # The default `strategy` block used to extract version information when
+        # a `strategy` block isn't provided.
+        DEFAULT_BLOCK = proc do |json, regex|
+          json["versions"]&.map do |version|
+            next if version["yanked"] == true
+            next unless (match = version["num"]&.match(regex))
+
+            match[1]
+          end
+        end.freeze
+
+        # The `Regexp` used to determine if the strategy applies to the URL.
+        URL_MATCH_REGEX = %r{
+          ^https?://static\.crates\.io/crates
+          /(?<package>[^/]+) # The name of the package
+          /.+\.crate # The crate filename
+        }ix
+
+        # Whether the strategy can be applied to the provided URL.
+        #
+        # @param url [String] the URL to match against
+        # @return [Boolean]
+        sig { params(url: String).returns(T::Boolean) }
+        def self.match?(url)
+          URL_MATCH_REGEX.match?(url)
+        end
+
+        # Extracts information from a provided URL and uses it to generate
+        # various input values used by the strategy to check for new versions.
+        #
+        # @param url [String] the URL used to generate values
+        # @return [Hash]
+        sig { params(url: String).returns(T::Hash[Symbol, T.untyped]) }
+        def self.generate_input_values(url)
+          values = {}
+          return values unless (match = url.match(URL_MATCH_REGEX))
+
+          values[:url] = "https://crates.io/api/v1/crates/#{match[:package]}/versions"
```

Comment:
```suggestion
          return values unless (package = url[URL_MATCH_REGEX, :package])

          values[:url] = "https://crates.io/api/v1/crates/#{package}/versions"
```

Reply:
As with the previous comment, this is a neat pattern but we use the `#match` approach in other strategies' `#generate_input_values` methods and I think there's value in consistency.

The `#match` approach executes the regex once regardless of the number of capture groups, so there's no additional overhead if we ever update this method to work with more than one capture group (and it's easy to do so). If we use the `url[URL_MATCH_REGEX, :package]` approach, we would be executing the regex each time we fetch a capture group.

The square-bracket pattern may also end up being more verbose when working with more than one capture group, since we would have to check for a good value from each capture group instead of doing one early `return` if the regex doesn't match. As above, the suggestion is tidier in this specific context but I would prefer to maintain consistency across strategies (and `strategy` blocks).

Reply:
> executes the regex once regardless of the number of capture groups

I think we should just use the simpler approach depending on the use case. It is easy enough to switch to a single `match` once we actually need multiple capture groups.